utils/config_stores.py

```diff
@@ -0,0 +1,226 @@
+# std
+import glob
+import imp
+import inspect
+import itertools
+import logging
+import simplejson as json
+from os import path
+
+# 3p
+from consul import Consul
+from etcd import EtcdKeyNotFound
+from etcd import Client as etcd_client
+from urllib3.exceptions import TimeoutError
+
+
+log = logging.getLogger(__name__)
+
+DEFAULT_ETCD_HOST = '127.0.0.1'
+DEFAULT_ETCD_PORT = 4001
+DEFAULT_ETCD_PROTOCOL = 'http'
+DEFAULT_RECO = True
+DEFAULT_TIMEOUT = 5
+SD_TEMPLATE_DIR = '/datadog/check_configs'
+
+DEFAULT_CONSUL_HOST = '127.0.0.1'
+DEFAULT_CONSUL_PORT = 8500
+DEFAULT_CONSUL_TOKEN = None
+DEFAULT_CONSUL_SCHEME = 'http'
+DEFAULT_CONSUL_CONSISTENCY = 'default'
+DEFAULT_CONSUL_DATACENTER = None
+DEFAULT_CONSUL_VERIFY = True
+
+AUTO_CONF_IMAGES = {
+    # image_name: [check_name, class_name]
+    'redis': ['redisdb', 'Redis'],
+    'nginx': ['nginx', 'Nginx'],
+    'consul': ['consul', 'ConsulCheck'],
+    'elasticsearch': ['elastic', 'ESCheck'],
+}
+
+
+class KeyNotFound(Exception):
+    pass
+
+
+class ConfigStore(object):
+    """Singleton for config stores"""
+    _instance = None
+
+    def __new__(cls, *args, **kwargs):
+        if cls._instance is None:
+            agentConfig = kwargs.get('agentConfig', {})
+            if agentConfig.get('sd_config_backend') == 'etcd':
+                cls._instance = object.__new__(EtcdStore, agentConfig)
+            elif agentConfig.get('sd_config_backend') == 'consul':
+                cls._instance = object.__new__(ConsulStore, agentConfig)
+        return cls._instance
+
+    def __init__(self, agentConfig):
+        self.client = None
+        self.agentConfig = agentConfig
+        self.settings = self._extract_settings(agentConfig)
+        self.client = self.get_client()
+        self.sd_template_dir = agentConfig.get('sd_template_dir')
+
+    def _extract_settings(self, config):
+        raise NotImplementedError()
+
+    def get_client(self, reset=False):
+        raise NotImplementedError()
+
+    def client_read(self, path, **kwargs):
+        raise NotImplementedError()
+
+    def _get_auto_config(self, image_name):
+        for key in AUTO_CONF_IMAGES:
+            if key == image_name:
+                check_name, class_name = AUTO_CONF_IMAGES[key]
+                check = self.get_check_class(check_name, class_name)
+                if check is None:
+                    log.info("Could not find an auto configuration template for %s."
+                             " Leaving it unconfigured." % image_name)
+                    return None
+                auto_conf = check.get_auto_config()
+                # stringify the dict to be consistent with what comes from the config stores
+                init_config_tpl = json.dumps(auto_conf.get('init_config'))
+                instance_tpl = json.dumps(auto_conf.get('instance'))
+                return [check_name, init_config_tpl, instance_tpl]
+        return None
+
+    def get_check_class(self, check_name, class_name):
```

Comment:
@remh I think so, but didn't want to do it in the same PR to avoid distracting the reviewer with a revamp of `load_check_directory`. I can do it anyway, or submit another PR later. What do you think?